File: SourceAFIS/Matching/MatchAnalysis.cs
using System;
using System.Collections.Generic;
using System.Text;
using SourceAFIS.General;

namespace SourceAFIS.Matching
{
    public sealed class MatchAnalysis
    {
        const int MinSupportingEdges = 1;
        const float DistanceErrorFlatness = 0.69f;
        const float AngleErrorFlatness = 0.27f;

        public int PairCount;
        public int CorrectTypeCount;
        public int SupportedCount;
        public float PairFraction;
        public int EdgeCount;
        public int DistanceErrorSum;
        public int AngleErrorSum;

        public void Analyze(MinutiaPairing pairing, FingerprintTemplate probe, FingerprintTemplate candidate)
        {
            var innerDistanceRadius = Convert.ToInt32(DistanceErrorFlatness * EdgeLookup.MaxDistanceError);
            var innerAngleRadius = Convert.ToInt32(AngleErrorFlatness * EdgeLookup.MaxAngleError);

            PairCount = pairing.Count;

            EdgeCount = 0;
            SupportedCount = 0;
            CorrectTypeCount = 0;
            DistanceErrorSum = 0;
            AngleErrorSum = 0;

            for (int i = 0; i < PairCount; ++i)
            {
                PairInfo pair = pairing.GetPair(i);
                if (pair.SupportingEdges >= MinSupportingEdges)
                    ++SupportedCount;
                EdgeCount += pair.SupportingEdges + 1;
                if (probe.Minutiae[pair.Pair.Probe].Type == candidate.Minutiae[pair.Pair.Candidate].Type)
                    ++CorrectTypeCount;
                if (i > 0)
                {
                    var probeEdge = EdgeConstructor.Construct(probe, pair.Reference.Probe, pair.Pair.Probe);
                    var candidateEdge = EdgeConstructor.Construct(candidate, pair.Reference.Candidate, pair.Pair.Candidate);
                    DistanceErrorSum += Math.Abs(probeEdge.Length - candidateEdge.Length);
                    AngleErrorSum += Math.Max(innerDistanceRadius, Angle.Distance(probeEdge.ReferenceAngle, candidateEdge.ReferenceAngle));
                    AngleErrorSum += Math.Max(innerAngleRadius, Angle.Distance(probeEdge.NeighborAngle, candidateEdge.NeighborAngle));
                }
            }

            float probeFraction = PairCount / (float)probe.Minutiae.Count;
            float candidateFraction = PairCount / (float)candidate.Minutiae.Count;
            PairFraction = (probeFraction + candidateFraction) / 2;
        }
    }
}

File: java/sourceafis/src/sourceafis/matching/MatchAnalysis.java
package sourceafis.matching;

import sourceafis.general.Angle;
import sourceafis.matching.minutia.EdgeLookup;
import sourceafis.matching.minutia.EdgeShape;
import sourceafis.matching.minutia.MinutiaPairing;
import sourceafis.matching.minutia.PairInfo;
import sourceafis.meta.Nested;
import sourceafis.meta.Parameter;
import sourceafis.templates.Template;
import sourceafis.matching.minutia.EdgeConstructor; 
public  class MatchAnalysis
{
    @Parameter(lower=0 , upper= 5)
    public int MinSupportingEdges = 1;
    @Parameter
    public float DistanceErrorFlatness = 0.69f;
    @Parameter
    public float AngleErrorFlatness = 0.27f;
    @Nested
    public EdgeConstructor edgeConstructor = new EdgeConstructor();

    public int maxDistanceError;
    public byte maxAngleError;
    public int pairCount;
    public int correctTypeCount;
    public int supportedCount;
    public float pairFraction;
    public int edgeCount;
    public int distanceErrorSum;
    public int angleErrorSum;
   
    public void analyze(MinutiaPairing pairing,EdgeLookup lookup,Template probe, Template candidate)
    {
    	  maxDistanceError = lookup.MaxDistanceError;
          maxAngleError = lookup.MaxAngleError;
          int innerDistanceRadius = Math.round(DistanceErrorFlatness * maxDistanceError);
          int innerAngleRadius = Math.round(AngleErrorFlatness * maxAngleError);

          pairCount = pairing.getCount();

          edgeCount = 0;
          supportedCount = 0;
          correctTypeCount = 0;
          distanceErrorSum = 0;
          angleErrorSum = 0;

        pairCount = pairing.getCount();
        edgeCount = 0;
        supportedCount = 0;
        correctTypeCount = 0;
        
        for (int i = 0; i < pairCount; ++i)
        {
            PairInfo pair = pairing.getPair(i);
            if (pair.supportingEdges >= MinSupportingEdges)
                ++supportedCount;
            edgeCount += pair.supportingEdges + 1;
            if (probe.minutiae[pair.pair.probe].Type == candidate.minutiae[pair.pair.candidate].Type)
                ++correctTypeCount;
            if (i > 0)
            {
                EdgeShape probeEdge = edgeConstructor.Construct(probe, pair.reference.probe, pair.pair.probe);
                EdgeShape candidateEdge = edgeConstructor.Construct(candidate, pair.reference.candidate, pair.pair.candidate);
                distanceErrorSum += Math.abs(probeEdge.length - candidateEdge.length);
                angleErrorSum += Math.max(innerDistanceRadius, Angle.Distance(probeEdge.referenceAngle, candidateEdge.referenceAngle) & 0xFF);
                angleErrorSum += Math.max(innerAngleRadius, Angle.Distance(probeEdge.neighborAngle, candidateEdge.neighborAngle) & 0xFF);
            }
            
        }
        float probeFraction = pairCount / (float)probe.minutiae.length;
        float candidateFraction = pairCount / (float)candidate.minutiae.length;
        pairFraction = (probeFraction + candidateFraction) / 2;
       
      
    }
}